### PR TITLE
Add WGSL standard utility package

### LINKED ISSUE #92: Design WGSL utilities library for reusable shader functions
## Summary

Design a reusable WGSL utilities library for VGPU that shader authors can import selectively through the existing `@vgpu/wgsl` resolver, while preserving native WGSL ergonomics, predictable emitted code, and strong testability.

This issue is **design/research only**. It proposes package shape, scope, testing, and licensing guardrails for a small, composable library of reusable WGSL functions such as remapping helpers, color conversions, and sampling/math primitives. **No implementation should start until maintainers approve the design.**

---

## Why this exists / agentic DX evidence and relation to #81/current workflow

Today VGPU already supports WGSL module graphs, import/export, tree-shaken inlining, validation, minification, and loader integration. What it does **not** ship yet is a vetted library of reusable shader helpers.

That gap shows up quickly in both human-written and agent-generated shader workflows:

- authors repeatedly rewrite the same helper functions across examples/apps
- agents tend to inline ad hoc shader snippets instead of importing known-good primitives
- repeated copy/paste makes provenance, testing, and bug-fixing harder
- teams lose the benefits of graph-level dedupe and centralized validation when every project carries slightly different remap/noise/color helpers

A small standard utility set would improve DX without inventing a new shader language:

- easier prompting and composition for agentic workflows
- less boilerplate in demos/examples/apps
- shared semantics and naming for common helpers
- centralized tests and licensing review
- better reuse of the resolver/import system already built in `@vgpu/wgsl`

This should relate to #81/current WGSL workflow as an extension of the existing import-based authoring model, not a replacement for it. The goal is to make common utilities easy to import while keeping authored shaders as normal WGSL modules.

---

## Existing VGPU WGSL package state and fit

Current repository/package state already provides almost all of the infrastructure needed:

- repo: pnpm monorepo, ESM-only, Node `>=22 <23`, TypeScript `NodeNext`
- package exists: `@vgpu/wgsl`
- `compile(wgsl: string)` is the plain-string entry point and explicitly rejects runtime `import`
- `@vgpu/wgsl/runtime` exports `resolveShader(...)`
- `resolveShader()` already returns:
  - `wgsl`
  - `deps`
  - `cacheKey`
  - `ast`
  - `sourceMap`
  - `diagnostics`
  - `reflection`
- existing resolver behavior already supports:
  - WGSL import syntax
  - relative imports
  - `@/` root imports
  - `packageMap`
  - npm package `exports`, including wildcard/conditional exports
  - virtual modules
  - graph-level tree shaking of reachable exports only
  - diamond-graph dedupe
  - stable path-keyed mangling
  - minification
  - validation
  - reflection
- loaders already exist:
  - `@vgpu/wgsl/loader-vite`
  - `@vgpu/wgsl/loader-webpack`
- TypeScript ambient types already exist: `@vgpu/wgsl/wgsl-types`
- tests already cover imports, mangler, minify, reflection, deps, loaders, package exports, diagnostics, cache keys, cycles, etc.

That means this proposal should **not** add a new include/macro system. The natural fit is a content package consumed through the current resolver.

Relevant constraints and repo conventions:

- ESM only
- `sideEffects: false`
- no runtime dependencies
- vitest-based tests
- GPU tests must remain gated by `VGPU_DOCKER_TEST=1`
- docs are allowlisted via `docs/allowlist.txt` with colocated docs and generated manifest flow
- bundle budgets matter:
  - `@vgpu/wgsl` main: `25KB gzip`
  - runtime/loader: `200KB`
- active PRs `#89`, `#90`, `#91` do not touch `packages/wgsl` source (`#89`/`#90` only regenerate `packages/vgpu/lib/generated/docs-manifest.generated.js`; `#91` is docs-only)
- release PR `#80` (v0.0.6) touches `packages/wgsl/package.json` only for a version bump (`0.0.5` → `0.0.6`) and regenerates the docs manifest; a new sibling package introduces no real conflict surface
- net effect: this work can land as a new package with effectively zero overlap against current open PRs

Suggested labels:

- `wgsl`
- `enhancement`
- `architecture`
- maybe `research` if maintainers want to emphasize design-track status

---

## State-of-the-art survey summary with links/citations

There is strong precedent for **modular composition** rather than shipping a giant monolithic shader stdlib.

### Precedents

- **glslify**: small npm-distributed shader modules with explicit imports  
  https://github.com/glslify/glslify
- **three.js ShaderChunk**: reusable chunks/includes, but more engine-managed and less native than desired here  
  https://github.com/mrdoob/three.js/tree/dev/src/renderers/shaders/ShaderChunk
- **three.js Nodes**: higher-level graph abstraction showing demand for reusable shading building blocks  
  https://threejs.org/docs/#api/en/nodes/Nodes
- **Babylon.js shader includes**: practical helper/includes library in a large engine  
  https://github.com/BabylonJS/Babylon.js/tree/master/packages/dev/core/src/Shaders/ShadersInclude
- **Filament**: careful rationale around color, exposure, tonemapping, and physically based rendering tradeoffs  
  https://google.github.io/filament/Filament.html
- **WGSL spec**: native language rules and constraints this library must stay within  
  https://www.w3.org/TR/WGSL/
- **webgpu-samples**: evidence that many helpers recur in practice across WebGPU examples  
  https://webgpu.github.io/webgpu-samples/
- **ashima/webgl-noise**: permissively licensed noise precedent often referenced by shader authors  
  https://github.com/ashima/webgl-noise
- **hughsk/glsl-noise**: modular GLSL noise packaging precedent  
  https://github.com/hughsk/glsl-noise
- **Shadertoy**: huge source of ideas, but dangerous for direct reuse because many snippets lack clear reusable licensing  
  https://www.shadertoy.com/

### Survey conclusions

- The strongest precedent is **many small importable modules**, not one large stdlib blob.
- Native-language composition ages better than custom macro/include DSLs.
- Common reusable categories repeatedly show up across engines and sample code:
  - math/safety
  - random/hash
  - sampling patterns
  - color space / exposure / tone mapping
  - noise/fbm
  - SDF helpers
  - easing/screen/camera helpers
- For a VGPU first pass, the safest path is **small, well-tested, provenance-clear primitives**.
- More opinionated shading content (advanced tonemaps, procedural noise families, SDF packs) should come later, if at all.

---

### Recommendation

Create a **separate content package**:

- **recommended:** `@vgpu/wgsl-std`

with domain subpath exports such as:

- `@vgpu/wgsl-std/math`
- `@vgpu/wgsl-std/color`
- `@vgpu/wgsl-std/sampling`
- `@vgpu/wgsl-std/hash`
- later maybe `@vgpu/wgsl-std/noise`, `@vgpu/wgsl-std/sdf`

This package would ship `.wgsl` source modules and docs, consumed through the existing `@vgpu/wgsl/runtime` resolver and the existing loaders.

### Why a separate package is the best fit

- decouples utility-content cadence from compiler/runtime cadence
- keeps `@vgpu/wgsl` focused on compilation/resolution infrastructure
- reduces risk to existing bundle budgets for compiler/runtime paths
- makes scope discipline easier: content package can grow conservatively without bloating core resolver code
- aligns with npm exports and current resolver/packageMap behavior already supported

### Critical content guardrail

Exported WGSL in the utility package should be **pure content only**:

Allowed exports:

- `fn`
- `const`
- `struct`
- `alias`

Disallowed exports:

- `var<...>`
- `override`

Reason: utilities must not secretly inject bindings, uniforms, samplers, textures, pipeline constants, or other hidden shader state.

### Additional API/content expectations

- import only what is needed
- no monolithic “include everything” chunk by default
- stable names and predictable output
- small composable functions preferred over giant helper files
- docs should state precision/range/performance caveats
- utilities should remain valid native WGSL, not pseudo-WGSL with preprocessing requirements

#### A) Put helpers directly inside `@vgpu/wgsl`

Pros:

- one package to discover
- possibly simpler marketing/docs

Cons:

- couples library content to compiler/runtime release cadence
- higher risk of scope creep and budget pressure
- mixes infrastructure concerns with authored shader content

**Verdict:** not recommended.

#### B) Use a nested subpath like `@vgpu/wgsl/std`

Pros:

- discoverable under the existing WGSL package name
- may feel “official”

Cons:

- still couples content and compiler package ownership/versioning
- harder to reason about bundle budget and package boundaries
- more likely to make the core package feel like a stdlib host

**Verdict:** viable, but weaker than a separate package.

#### C) Build a custom include/macro system

Pros:

- none compelling relative to current resolver capabilities

Cons:

- duplicates existing import graph features
- introduces another authoring model to learn/debug
- risks moving away from native WGSL semantics and testability

**Verdict:** explicitly avoid.

---

### Proposed final usage

```wgsl
import { saturate, remap, safe_normalize } from "@vgpu/wgsl-std/math";
import { srgb_to_linear, linear_to_srgb, luminance, apply_exposure } from "@vgpu/wgsl-std/color";
import { concentric_disk, vogel_disk } from "@vgpu/wgsl-std/sampling";

struct VSOut {
  @builtin(position) position : vec4<f32>,
  @location(0) uv : vec2<f32>,
};

@fragment
fn fs_main(in: VSOut) -> @location(0) vec4<f32> {
  let uv = in.uv;
  let t = saturate(remap(uv.x, 0.1, 0.9, 0.0, 1.0));
  let n = safe_normalize(vec3(uv - 0.5, 0.25));
  let color = vec3(t, luminance(vec3(n.xy, 0.5)), 1.0 - t);
  let exposed = apply_exposure(color, 1.0);
  return vec4(linear_to_srgb(exposed), 1.0);
}
```

### Before: manual copy/paste helpers in each shader

```wgsl
fn remap(v: f32, in_min: f32, in_max: f32, out_min: f32, out_max: f32) -> f32 {
  return out_min + (v - in_min) * (out_max - out_min) / (in_max - in_min);
}

fn saturate(x: f32) -> f32 {
  return clamp(x, 0.0, 1.0);
}

fn linear_to_srgb(c: vec3<f32>) -> vec3<f32> {
  let cutoff = vec3<f32>(0.0031308);
  let low = c * 12.92;
  let high = 1.055 * pow(c, vec3<f32>(1.0 / 2.4)) - vec3<f32>(0.055);
  return select(high, low, c <= cutoff);
}

@fragment
fn fs_main() -> @location(0) vec4<f32> {
  let t = saturate(remap(0.3, 0.0, 1.0, 0.0, 1.0));
  return vec4(linear_to_srgb(vec3(t)), 1.0);
}
```

### After: imported, shared, testable helpers

```wgsl
import { remap, saturate } from "@vgpu/wgsl-std/math";
import { linear_to_srgb } from "@vgpu/wgsl-std/color";

@fragment
fn fs_main() -> @location(0) vec4<f32> {
  let t = saturate(remap(0.3, 0.0, 1.0, 0.0, 1.0));
  return vec4(linear_to_srgb(vec3(t)), 1.0);
}
```

### Native/manual override remains possible

Authors should still be free to write bespoke WGSL manually:

```wgsl
fn my_filmic_curve(x: vec3<f32>) -> vec3<f32> {
  // app-specific choice; not something stdlib must own
  return x / (x + vec3(1.0));
}
```

This proposal is about **reusable imports**, not forcing a standardized shading style.

---

#### `/math`

- `saturate` / `clamp01`
- `inverse_lerp`
- `remap` / `value_remap` (pick one canonical name; alias only if justified)
- `safe_normalize`
- `rotate2d` and/or small rotation helpers

#### `/color`

- `linear_to_srgb`
- `srgb_to_linear`
- `luminance`
- `apply_exposure`
- maybe `reinhard_tonemap` **only if maintainers want a very small tonemap foothold**

#### `/hash` or `/random`

- `hash_u32`
- `hash_vec2`
- `random01`

#### `/sampling`

- `concentric_disk`
- `vogel_disk`

### v1.1 candidates

- advanced tonemaps:
  - ACES approximation
  - Hable / Uncharted-style filmic curves
- noise family **only if provenance is explicit and implementation is original or cleanly adapted from compatible sources**:
  - value noise
  - Perlin-like noise
  - fbm helpers
- simple SDF 2D primitives/ops **only if formulas are implemented originally or sources are clearly compatible**
- easing helpers
- screen helpers

### later / maybe never

- large PBR/material helper suites
- opinionated BRDF libraries
- engine-specific camera/layout assumptions
- macro-driven meta-programming
- giant all-in-one shader chunk packs

---

### 1) Math helpers (`@vgpu/wgsl-std/math`)

**Purpose**

Small scalar/vector helpers that are repeatedly reimplemented in shaders and are easy to specify, test, and document.

**Candidate WGSL names**

- `saturate(x)` or `clamp01(x)`
- `inverse_lerp(a, b, v)`
- `remap(v, in_min, in_max, out_min, out_max)`
- `value_remap(...)` only if aliasing is worth supporting
- `safe_normalize(v)`
- `rotate2d(v, radians)` and/or `rotation2d(radians)` returning `mat2x2<f32>`

**Performance notes**

- `saturate` should compile to trivial clamp logic; no issue.
- `inverse_lerp`/`remap` need documented behavior near zero-width ranges; avoid undefined helper semantics.
- `safe_normalize` trades one branch/guard for avoiding NaNs/infs on tiny vectors.
- `rotate2d` uses trig; should be documented as fine for occasional coordinate transforms, not free in hot loops.

**Test approach**

- Tier 0 string/resolution snapshots for imports and emitted WGSL
- Tier 1 validation harnesses for scalar/vector overloads or named variants
- Tier 2 tiny GPU numeric tests against CPU references for:
  - remap endpoints/midpoints
  - safe-normalize zero and nonzero vectors
  - rotate2d known angles

### 2) Color basics (`@vgpu/wgsl-std/color`)

**Purpose**

Provide foundational color-space and exposure helpers that are broadly useful and have clear semantics.

**Candidate WGSL names**

- `linear_to_srgb(c)`
- `srgb_to_linear(c)`
- `luminance(c)`
- `apply_exposure(c, exposure)`
- maybe `reinhard_tonemap(c)` if accepted for v1

**Performance notes**

- sRGB conversions require piecewise logic and `pow`; correctness matters more than shaving a line of code.
- `luminance` is cheap and broadly useful.
- `apply_exposure` is trivial and should stay narrowly defined.
- Tonemap inclusion should be conservative because it starts to encode rendering policy, not just utility math.

**Test approach**

- Tier 0 snapshots for emitted code and imports
- Tier 1 validation with trivial fragment harnesses
- Tier 2 tiny GPU numeric comparisons vs CPU references for:
  - known sRGB/linear boundary values
  - luminance of canonical colors
  - exposure scaling
  - Reinhard mapping if included

### 3) Sampling helpers (`@vgpu/wgsl-std/sampling`) — optional stretch

**Purpose**

Reusable coordinate-generation helpers for common scatter/distribution patterns.

**Candidate WGSL names**

- `concentric_disk(uv)`
- `vogel_disk(index, count, angle_offset)` or similarly explicit signature

**Performance notes**

- document coordinate/range conventions carefully
- `vogel_disk` is useful but should avoid hidden assumptions about sample count ordering or normalization
- good candidates when used for DOF/soft shadow/sample distributions, but still more specialized than math/color

**Test approach**

- Tier 0 import/golden tests
- Tier 1 validation harnesses
- Tier 2 tiny 4x4 or 8x8 tests only where neighborhood/distribution matters, with CPU reference checks for point locations and radius monotonicity where applicable

### 4) Hash/random helpers (`@vgpu/wgsl-std/hash` or `/random`) — optional stretch

**Purpose**

Provide deterministic, tiny pseudo-random/hash utilities for shader-local sampling/jitter, not cryptography.

**Candidate WGSL names**

- `hash_u32(x)`
- `hash_vec2(v)`
- `random01(seed)`

**Performance notes**

- must document deterministic/non-cryptographic nature
- avoid overpromising statistical quality
- keep API narrow and signatures explicit

**Test approach**

- Tier 0 determinism snapshots and import tests
- Tier 1 validation harnesses
- Tier 2 tiny GPU tests only for deterministic equality against CPU reference, not big image-based “looks random” testing

---

## Strong cheap testing strategy with CI budgets

Testing needs to stay strong **and** cheap.

### Tier 0 — fast required TS/unit tests

Budget:

- target `<5s` local
- target `<15s` CI for the package

Coverage:

- generated WGSL string golden tests
- import/export resolution tests
- tree-shaking of only referenced exports
- minify lexical/semantic determinism
- cache-key determinism
- small snapshots for emitted WGSL, AST shape, and reflection shape
- graph determinism across direct import, re-export, and diamond graphs

These should be the bulk of the test suite.

### Tier 1 — compile-only WGSL validation

Budget:

- roughly `5–20s`

Coverage:

- `resolveShader({ validate: true })` over generated utility-entry harnesses
- diagnostics/source mapping checks on authored modules
- minified-output validation checks (i.e. `minify: true` whitespace/comment compaction) to confirm validation still reports against authored modules

This verifies utilities are valid WGSL across the real resolver pipeline without requiring large GPU integration tests.

### Tier 2 — tiny numeric GPU tests, selective and gated

Budget:

- roughly `10–60s`
- only when `VGPU_DOCKER_TEST=1`

Coverage guidelines:

- 1x1 readback tests for scalar/color helpers
- 4x4 UV/sampling tests for small coordinate-distribution helpers
- 8x8 only when neighborhood behavior truly matters
- compare GPU outputs to CPU references within epsilon

Explicit principle:

- **render tests must stay tiny and cheap**
- **do not** use large golden images for this library

### Additional testing principles

- every pure function should have a CPU reference where practical
- snapshots should be curated and stable, not massive dumps
- test determinism across:
  - minify modes
  - import graph shapes
  - cache-key generation
- imported helper output should remain predictable even when graph structure changes

---

## Licensing/attribution policy

This is critical.

Shader snippets are copyrighted code. The project should adopt an explicit provenance policy before implementation.

### Policy proposal

- prefer original WGSL implementations derived from formulas, standards, papers, or permissive references
- do **not** copy code from sources with unclear licensing
- do **not** copy GPL/AGPL/LGPL shader snippets into the library
- do **not** copy from Shadertoy unless the specific source has explicit compatible licensing and attribution requirements are satisfied
- do **not** copy Stack Overflow/blog snippets without clear reuse rights

### Acceptable reference classes

- formulas/specs/papers re-expressed originally in WGSL
- permissively licensed references with preserved notices where needed
- examples from clearly compatible sources such as:
  - three.js (MIT)
  - Babylon.js (Apache-2.0)
  - Filament docs/code where applicable under Apache-2.0 terms
  - ashima/webgl-noise (MIT)
  - hughsk/glsl-noise (MIT)

### Required process for any adopted utility

- provenance record per utility/module
- source comments or metadata for directly derived files
- SPDX/license/source comments when code is directly adapted
- preserve upstream copyright/license notices in copied or adapted source files when required
- include the upstream license text in repo/package licensing materials when required
- if an Apache-2.0 source includes a `NOTICE` file, carry forward the relevant notices in project notices/docs as required
- mark modified files prominently when adapting Apache-2.0-licensed code
- docs should mention precision/perf semantics and provenance where relevant

If provenance is uncertain, the default answer should be **do not include it**.

---

## Non-goals

- building a giant WGSL standard library
- introducing macros, templating, or a new include system
- hiding bindings/uniforms/overrides inside “helper” imports
- shipping opinionated material/PBR frameworks in v1
- guaranteeing every popular shader trick will become part of VGPU
- large visual regression suites or expensive golden-image pipelines for utility validation
- replacing hand-authored WGSL with a higher-level DSL

---

## Acceptance criteria

- [ ] Maintainers agree this remains a **design-only** issue until explicit approval
- [ ] A package direction is chosen (`@vgpu/wgsl-std` recommended, alternatives documented)
- [ ] Scope is intentionally narrow for v1: at least `/math` and `/color`; sampling/hash only if approved
- [ ] Utility exports are restricted to pure content (`fn`, `const`, `struct`, `alias`) with no exported bindings/overrides/state
- [ ] Import model reuses the existing `@vgpu/wgsl` resolver/package exports; no new macro/include system
- [ ] Each accepted utility has documented semantics, range/precision notes, and test strategy
- [ ] A provenance/licensing policy is agreed before code lands
- [ ] Tests are tiered with explicit CI/runtime budgets
- [ ] Docs plan fits existing colocated docs + allowlist/manifest conventions
- [ ] Bundle-size impact is understood and acceptable for the chosen package boundary
- [ ] Issue outcome clearly identifies v1, v1.1, and non-goals

---

## Open questions

1. **Package name/boundary:** should this be `@vgpu/wgsl-std`, `@vgpu/wgsl-utils`, or `@vgpu/wgsl/std`?
2. **Canonical naming:** should v1 expose `remap`, `value_remap`, or both?
3. **Scalar/vector overload style:** prefer separate typed helpers, or one name with multiple WGSL overloads where practical?
4. **Tonemap scope:** should `reinhard_tonemap` be included in v1, or should all tonemapping wait for the broader PBR/rendering discussion in #41?
5. **Stretch scope:** should `/sampling` and `/hash` ship in v1 if they meet provenance/tests, or should v1 stay only `/math` + `/color`?
6. **Docs shape:** should each utility get a colocated docs page, or should docs be grouped by module plus generated manifest entries?
7. **Provenance tracking format:** comments in WGSL files only, or a structured manifest in addition?
8. **Reflection expectations:** do we want any extra metadata conventions for stdlib utilities, or is plain emitted WGSL enough?
9. **Versioning cadence:** should the content package release independently from `@vgpu/wgsl`, even if maintained in the same monorepo?
10. **Adoption bar:** is “two real uses + portability + testability + provenance” the right threshold for accepting new utilities?

---

## Explicit statement: design issue only; implementation waits for maintainer approval

This issue is intentionally scoped to **design/research**.

It should produce agreement on:

- package boundary
- v1 scope
- API/content guardrails
- test budgets
- licensing/provenance policy

**No implementation work should begin from this issue alone.** Actual code, package creation, docs wiring, or utility authoring should wait for explicit maintainer approval on the proposed design.

### PR DESCRIPTION
**Pull request by matiasngf**
Closes #92.

## Summary

This adds the new `@vgpu/wgsl-std` package: raw WGSL utility modules that resolve through `@vgpu/wgsl` without adding any root WGSL export or hidden runtime resources.

Final API/package shape:

- Package: `@vgpu/wgsl-std`.
- Imports are explicit subpaths only:
  - `@vgpu/wgsl-std/math`
  - `@vgpu/wgsl-std/color`
  - `@vgpu/wgsl-std/sampling`
  - `@vgpu/wgsl-std/constants`
- No root WGSL export: importing `@vgpu/wgsl-std` as WGSL intentionally fails.
- Shipped WGSL is pure `fn`/`const` declarations only.
- No hidden bindings, uniforms, textures, samplers, storage, overrides, or runtime state.

## Utility catalog shipped

Constants:

- `pi`, `tau`, `halfPi`, `quarterPi`, `invPi`, `invTau`, `goldenRatio`, `goldenAngle`.

Math:

- `saturate`
- `clamp01`
- `inverseLerp`
- `remap`
- `safeNormalize2`
- `safeNormalize3`
- `safeNormalize4`
- `rotate2d`

Color:

- `srgbToLinear`
- `srgbToLinear3`
- `srgbToLinear4`
- `linearToSrgb`
- `linearToSrgb3`
- `linearToSrgb4`
- `luminance`
- `applyExposure`

Sampling:

- `goldenAngle`
- `vogelDisk`
- `radicalInverseVdc`
- `hammersley2d`

## Intentional deferrals

Sampling helpers are shipped in this PR; they are **not** deferred.

The following remain intentionally out of scope for this first standard WGSL package:

- PBR helpers.
- PBR/filmic tonemaps.
- BRDF/Fresnel/IBL helpers.
- Perlin/simplex/fbm/value noise.
- Risky shader hash/random snippets.
- `concentricDisk`, which is a good fast-follow candidate but needs dedicated API and edge-behavior review before shipping.

Declaration-level dead-code elimination / tree-shaking for larger WGSL modules is tracked separately in #98 after PR review follow-up confirmed the current resolver/loaders do not remove unused declarations inside a loaded module.

## Example: before and after

Before, users copied utility WGSL manually into shader strings:

```wgsl
const PI: f32 = 3.1415927;

fn saturate(x: f32) -> f32 {
  return clamp(x, 0.0, 1.0);
}

fn applyExposure(color: vec3f, exposure: f32) -> vec3f {
  return color * pow(2.0, exposure);
}
```

After, users import named utility modules through the existing WGSL resolver:

```wgsl
#import { pi, goldenAngle } from "@vgpu/wgsl-std/constants"
#import { saturate, safeNormalize3 } from "@vgpu/wgsl-std/math"
#import { srgbToLinear3, applyExposure } from "@vgpu/wgsl-std/color"
#import { vogelDisk, hammersley2d } from "@vgpu/wgsl-std/sampling"

fn shade(albedo_srgb: vec3f, normal: vec3f, sample_index: u32, sample_count: u32) -> vec3f {
  let albedo = srgbToLinear3(albedo_srgb);
  let n = safeNormalize3(normal);
  let disk = vogelDisk(sample_index, sample_count, goldenAngle + pi * 0.0);
  let h = hammersley2d(sample_index, sample_count);
  return applyExposure(albedo * saturate(n.z + disk.x * 0.0 + h.y * 0.0), 0.0);
}
```

## Testing and validation

Latest validation from `feature/issue-92-wgsl-utils` worktree after review follow-up:

- `pnpm exec vitest run packages/wgsl-std/tests --reporter=verbose`
  - Result: PASS — 6 files passed; 30 tests passed; 4 native-gated tests skipped.
- `pnpm typecheck`
  - Result: PASS — `tsc -b` completed successfully.
  - Environment warning only: `Unsupported engine: wanted: {"node":">=22 <23"} (current: {"node":"v20.20.0","pnpm":"9.15.4"})`.
- `pnpm build`
  - Result: PASS — `tsc -b` and WGSL type declaration copy completed successfully.
  - Same Node engine warning as above.
- `pnpm exec vitest run packages/wgsl/tests packages/wgsl-std/tests`
  - Result: PASS — 25 files passed, 1 file skipped; 168 tests passed; 18 tests skipped.

Additional review/QA validation reported before final polish:

- `pnpm exec vitest run packages/wgsl-std/tests --reporter=verbose` passed 27 tests with 4 native-gated skipped.
- npm pack dry-run was okay.
- Native/Naga/GPU validation is gated by `VGPU_DOCKER_TEST=1`; the current sandbox cannot run it because of the known native `GLIBC_2.38` limitation.

## Review cycle summary

Integrated slice PRs into this feature branch:

- #93 scaffold/resolver/purity — implementer `bd902fe0`, reviewers `b822d90f`, `b5a6fbf2`.
- #94 math — implementer `55357873`, reviewers `a7308804`, `ec54aa5f`.
- #95 color — implementer `bde06e31`, reviewers `e6d2ddbc`, `48290c3e`.
- #96 sampling — implementer `59f1f997`, reviewers `44c7d94e`, `bb6e31dc`.

Final reviews:

- QA/test debug-high `73c3966e-fe18-4bb8-b57c-e6c0196a6cda`: PASS/APPROVE, no blockers.
- Scope/API reason `8295d2e5-1ef8-4b59-83be-29e964073273`: APPROVE, nonblocking caveats only. Requested the root README package table polish.
- Licensing/provenance reason `d0d3a2a1-57e3-4e6b-87d8-60e36d89ded0`: APPROVE, no blockers.
- PR #97 user follow-up requested a math constants module and asked whether resolver/loaders perform dead-code elimination. This update adds `@vgpu/wgsl-std/constants` and opened #98 for declaration-level WGSL DCE design.

Phase-1 research/exploration:

- Explorer `640512fa`.
- Research `3d7a6cd2`.
- Design `f019e251`.
- Optional-scope reason `5451c176`.

## Licensing and provenance

The utilities are standard public formulas and original transcriptions for this package:

- Standard mathematical constants rounded to WGSL `f32` literals.
- sRGB transfer constants from IEC 61966-2-1.
- Rec.709 luminance coefficients.
- Vogel 1979 / golden-angle disk sampling.
- Van der Corput radical inverse and Hammersley standard low-discrepancy sequences.

No Shadertoy, StackOverflow, or similar snippet code was copied. Package metadata is MIT and no new dependencies were added.

## Performance and CI notes

- Tests are cheap CPU/reference/resolver/minify coverage.
- Docker/native validation remains explicitly gated.
- No new hidden resources or runtime state.
- No root export for `@vgpu/wgsl-std`; use explicit subpaths.
- Current resolver/loaders do not perform declaration-level DCE inside loaded modules; #98 tracks design for that feature. The new constants module is intentionally small.

## Issue wording note

The literal issue text for #92 was unavailable/private during agent execution. Please verify the exact #92 wording before merge. The shipped catalog reflects the approved execution scope and the user preference to avoid deferrals where scope/review allowed; sampling is included in this PR.

## Signed off by orchestrator

- Session: `6588949f-823e-46d0-98cd-19aa99c8f626`
- Date: `2026-05-30`

### CODE REVIEW

File: packages/wgsl-std/src/sampling/index.wgsl [hunk 1]
```diff
@@ -0,0 +1,31 @@
+export const goldenAngle: f32 = 2.3999631;
```
Comment (matiasngf) [inline packages/wgsl-std/src/sampling/index.wgsl:1]: woudl be nice to have a math constants module

File: packages/wgsl-std/src/math/index.wgsl [hunk 2]
```diff
@@ -0,0 +1,50 @@
+export fn saturate(value: f32) -> f32 {
```
Comment (matiasngf) [inline packages/wgsl-std/src/math/index.wgsl:1]: research if the vgpu wasm loader supports deleting dead code, otherwise modules might get too big. if not, create an issue to design the feature

File: packages/wgsl-std/src/sampling/index.wgsl (hunk 1 above)
Comment (matiasngf) [reply packages/wgsl-std/src/sampling/index.wgsl:1]: Added in 1a6c563: `@vgpu/wgsl-std/constants` now exports common pure WGSL math constants (`pi`, `tau`, `halfPi`, `quarterPi`, `invPi`, `invTau`, `goldenRatio`, `goldenAngle`) with package exports, docs, resolver tests, and purity coverage.

File: packages/wgsl-std/src/math/index.wgsl (hunk 2 above)
Comment (matiasngf) [reply packages/wgsl-std/src/math/index.wgsl:1]: Checked this while doing the follow-up. The current WGSL resolver/loaders load reachable modules and run whitespace/safe identifier minification, but they do not do declaration-level dead-code elimination inside a loaded module. I opened #98 to design that feature. For this PR, the added constants module is intentionally small and documented with the #98 note.
